airbyte-integrations/connectors/source-pinterest/source_pinterest/reports/reports.py

```diff
@@ -0,0 +1,171 @@
+            report_status, report_url = self._verify_report_status(report_info, stream_slice)
+            report_info.report_status = report_status
+            if report_status in {ReportStatus.DOES_NOT_EXIST, ReportStatus.EXPIRED, ReportStatus.FAILED, ReportStatus.CANCELLED}:
+                message = "Report generation failed."
+                raise ReportGenerationFailure(message)
+            elif report_status == ReportStatus.FINISHED:
+                try:
+                    report_info.metrics = self._fetch_report_data(report_url)
+                except requests.HTTPError as error:
+                    raise ReportGenerationFailure(error)
+
+        pending_report_status = [report_info for report_info in report_infos if report_info.report_status != ReportStatus.FINISHED]
+
+        if len(pending_report_status) > 0:
+            message = "Report generation in progress."
+            raise ReportGenerationInProgress(message)
+
+    def _incomplete_report_infos(self, report_infos):
+        """Return the report infos which are not yet finished."""
+        return [r for r in report_infos if r.report_status != ReportStatus.FINISHED]
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        """Parse the API response."""
+        yield response.json()
+
+    @backoff_max_tries
+    def _init_reports(self, init_reports) -> List[ReportInfo]:
+        """Initialize the reports and return them as a list."""
+        report_infos = []
+        for init_report in init_reports:
+            status = ReportInfo.parse_raw(json.dumps(init_report))
+            report_infos.append(
+                ReportInfo(
+                    token=status.token,
+                    report_status=ReportStatus.IN_PROGRESS,
+                    metrics=[],
+                )
+            )
+        self.logger.info("Initiated successfully.")
+        return report_infos
+
+    def _http_get(self, url, params=None, headers=None):
+        """Make a GET request to the given URL and return the response as a JSON."""
+        response = self._session.get(url, params=params, headers=headers)
+        response.raise_for_status()
+        return response.json()
+
+    def _verify_report_status(self, report: dict, stream_slice: Mapping[str, Any]) -> tuple:
+        """Verify the report status and return it along with the report URL."""
+        api_path = self._build_api_path(stream_slice["parent"]["id"])
+        response_data = self._http_get(
+            urljoin(self.url_base, api_path), params={"token": report.token}, headers=self.authenticator.get_auth_header()
+        )
+        try:
+            report_status = ReportStatusDetails.parse_raw(json.dumps(response_data))
+        except ValueError as error:
+            raise ReportStatusError(error)
+        return report_status.report_status, report_status.url
+
+    def _fetch_report_data(self, url: str) -> dict:
```

Comment:
perhaps we should wrap `_fetch_report_data` and `_verify_report_status` in streams to make it retriable and unify interfaces

Reply:
I understand that both `_fetch_report_data` and `_verify_report_status` may share some similarities in their nature, but it's important to acknowledge that they serve distinct purposes. Leaving them as they are would actually help maintain code clarity, as it allows each function to focus on its specific task without combining their functionalities.

Reply:
I didn't mean to wrap them in a single stream, rather two streams. But it's ok to leave it for later, when the oncall issue is resolved